onnxruntime/core/mlas/lib/q4_dq.cpp

```diff
@@ -638,6 +638,691 @@ struct BlockwiseQuantizer {
+     * @param dst_scales        [columns, ceil(rows / quant_block_size)]
+     * @param dst_zero_points   [columns, ceil(ceil(rows / quant_block_size) * qbits / 8)] in uint8_t.
+     * @param rows              number of src rows in qbits type.
+     * @param columns           number of src columns in qbits type.
+     * @param quant_block_size  number of elements quantized together
+     * @param thread_pool       thread pool for parallel processing
+     */
+    static void TransposeColumnWiseQuantized(
+        const uint8_t* src_weights,
+        const Tin* src_scales,
+        const uint8_t* src_zero_points,
+        uint8_t* dst_weights,
+        Tin* dst_scales,
+        uint8_t* dst_zero_points,
+        int32_t rows,
+        int32_t columns,
+        int32_t quant_block_size,
+        MLAS_THREADPOOL* thread_pool
+    )
+    {
+        // Must avoid multiple thread write to a single byte, which means the starting index
+        // of a thread block must be even. To achieve that, we need to customize the thread
+        // block size based on the parity of columns.
+        if (columns & 1) {
+            TransposeColumnWiseQuantizedPackUnaligned(
+                src_weights, src_scales, src_zero_points,
+                dst_weights, dst_scales, dst_zero_points,
+                rows, columns, quant_block_size, thread_pool
+            );
+        } else {
+            TransposeColumnWiseQuantizedPackAligned(
+                src_weights, src_scales, src_zero_points,
+                dst_weights, dst_scales, dst_zero_points,
+                rows, columns, quant_block_size, thread_pool
+            );
+        }
+    }
+
+private:
+    static void QuantizeColumnWisePackAligned(
+        const Tin* src,
+        Tin* scales,
+        uint8_t* zero_points,
+        uint8_t* dst,
+        int32_t rows,
+        int32_t columns,
+        int32_t quant_block_size,
+        MLAS_THREADPOOL* thread_pool
+    )
+    {
+        ORT_ENFORCE(columns % 2 == 0, "Columns must be multiple of 2.");
+        // Thread block is [quant_block_size, thread_blk_size]. thread_blk_size % 2 == 0.
+        const int32_t thread_blk_size = 64;
+        const auto num_row_thread_blk = (rows + quant_block_size - 1) / quant_block_size;
+        const auto num_col_thread_blk = (columns + thread_blk_size - 1) / thread_blk_size;
+        const auto num_thread_blk = num_row_thread_blk * num_col_thread_blk;
+        const auto minf = std::numeric_limits<float>::lowest();
+        const auto maxf = std::numeric_limits<float>::max();
+
+        MlasTryBatchParallel(
+            thread_pool, static_cast<ptrdiff_t>(num_thread_blk),
+            [&](ptrdiff_t thread_blk_idx) {
+                // !!warning!!: buffering the whole thread block
+                const int32_t buffer_size = 64;
+                ORT_ENFORCE(buffer_size == thread_blk_size, "buffer size must be equal to thread block size.");
```

Comment:
nit: use thread_blk_size?

Reply:
the lamda captures variables like `thread_blk_size` during runtime. Even though thread_blk_size is a const, if I do `buffer_size = thread_blk_size`, `buffer_size` becomes a runtime variable and cannot be used in `float x[buffer_size]`.

Reply:
you can declare 'thread_blk_size' with constexpr instead of const. it can be used in the lambda directly without capture.

Reply:
seems constexpr also failed the compilation. I guess the compiler requests the size in x[size] to be determined at compile time

Reply:
looks like MSVC has a different implementation of lambda const capture. Because of that, thread_blk_size in lambda is known at compile time. 
[ref1](https://www.reddit.com/r/cpp_questions/comments/5g4jx9/behavior_of_lambda_capture_concerning_constexpr/), [ref2](https://stackoverflow.com/questions/73768430/lambda-const-capture-by-value-bug-in-msvc)